central/notifications/datastore/internal/writer/writer_impl.go

```diff
@@ -0,0 +1,139 @@
+}
+
+func (c *writerImpl) read(id string) (*storage.Notification, bool) {
+	notification, found := c.buffer[id]
+	return notification, found
+}
+
+func (c *writerImpl) write(notification *storage.Notification) {
+	c.buffer[notification.GetId()] = notification
+}
+
+func (c *writerImpl) Upsert(ctx context.Context, notification *storage.Notification) error {
+	enrichedNotification := enrichNotificationWithDefaults(notification)
+	id := enrichedNotification.GetId()
+
+	var baseNotification *storage.Notification
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	notificationInBuffer, found := c.read(id)
+	// If a notification already exists in the buffer it is the most recent.
+	// We use it as a base to merge with the new notification.
+	if found {
+		baseNotification = notificationInBuffer
+	} else {
+		// If no notification is in the buffer, we try to fetch a notification
+		// from the database. If found, we use it as the base for the merge.
+		notificationInDB, found, err := c.store.Get(ctx, id)
+		if err != nil {
+			return errors.Wrap(err, "failed to query for existing record")
+		}
+		if found {
+			baseNotification = notificationInDB
+		}
+	}
+
+	// Merge notifications to up the occurrence and update the time stamps.
+	if baseNotification != nil {
+		enrichedNotification = mergeNotifications(baseNotification, enrichedNotification)
+	}
+
+	c.write(enrichedNotification)
+	return nil
+}
+
+func (c *writerImpl) Flush(ctx context.Context) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	notificationChunk := make([]*storage.Notification, 0, len(c.buffer))
+	for _, notification := range c.buffer {
+		notificationChunk = append(notificationChunk, notification)
+	}
+	c.buffer = make(map[string]*storage.Notification)
+	err := c.store.UpsertMany(ctx, notificationChunk)
+	if err != nil {
+		return errors.Wrap(err, "failed to upsert notification chunk")
+	}
+	return nil
+}
+
+func getNotificationID(notification *storage.Notification) string {
+	dedupKey := strings.Join([]string{
+		notification.GetDomain(),
+		notification.GetMessage(),
+		notification.GetResourceId(),
+		notification.GetResourceType(),
+		notification.GetType().String(),
+	}, ",")
+	return uuid.NewV5(rootNamespaceUUID, dedupKey).String()
+}
+
+func enrichNotificationWithDefaults(notification *storage.Notification) *storage.Notification {
+	if notification == nil {
+		return nil
+	}
+
+	enrichedNotification := notification.Clone()
+	enrichedNotification.Id = getNotificationID(notification)
+	if notification.GetNumOccurrences() == 0 {
+		enrichedNotification.NumOccurrences = 1
+	}
+	if notification.GetCreatedAt() == nil {
+		enrichedNotification.CreatedAt = protoconv.ConvertTimeToTimestamp(time.Now())
+	}
+	if notification.GetLastOccurredAt() == nil {
+		enrichedNotification.LastOccurredAt = protoconv.ConvertTimeToTimestamp(time.Now())
+	}
+	return enrichedNotification
+}
+
+func mergeNotifications(base *storage.Notification, new *storage.Notification) *storage.Notification {
```

Comment:
Since we seem to anyways assign the return value to the previous `new` value, why not modify the `new` notification in place and avoid the `Clone()` call, and explicitly call this out in the function description?

Basically, want to avoid calling `Clone()` here as this might have quite some high volume, and there's no strict requirement to do so (albeit, admittedly, the proto is quite small).